server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -303,11 +304,27 @@ public void finalizeRecovery(final long globalCheckpoint, ActionListener<Void> l
             // Persist the global checkpoint.
             indexShard.sync();
             indexShard.persistRetentionLeases();
+            if (hasUncommittedOperations()) {
+                indexShard.flush(new FlushRequest().force(true).waitIfOngoing(true));
+            }
             indexShard.finalizeRecovery();
             return null;
         });
     }
 
+    private boolean hasUncommittedOperations() {
+        if (indexShard.translogStats().getUncommittedOperations() == 0) {
+            return false;
+        }
+        // In peer recovery, we transfer history from primary to replica, thus we don't have to flush
+        // if all those uncommitted operations have baked into the existing Lucene index commit already.
+        final SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+            indexShard.commitStats().getUserData().entrySet());
+        return commitInfo.maxSeqNo != commitInfo.localCheckpoint
```

Comment:
I wonder if the condition above about the translog is sufficient. What situation is the condition here addressing that's not addressed by the above one?

Reply:
If a file-based occurs, the primary also sends its translog to replica. These operations are uncommitted on the replica even though they are baked into the commit already.  We need this condition to avoid flushing in this case to keep the syncId. I pushed https://github.com/elastic/elasticsearch/pull/41660/commits/07c3a7c03cc493b63ea6c7300fb4dbc1bbe501e9 to use another check.